quickwit-search/src/collector.rs

```diff
@@ -0,0 +1,376 @@
+use tantivy::fastfield::DynamicFastFieldReader;
+use tantivy::fastfield::FastFieldReader;
+use tantivy::DocId;
+use tantivy::Score;
+use tantivy::SegmentOrdinal;
+use tantivy::SegmentReader;
+
+/// Defines the way documents should be sorted.
+/// In case of a tie, the documents are ordered according to descending `(split_id, segment_ord, doc_id)`.
+#[derive(Clone, Debug)]
+pub enum SortBy {
+    /// Sort by a specific field.
+    SortByFastField {
+        field_name: String,
+        order: SortOrder,
+    },
+    /// If undefined, we sort by `DocId`
+    Undefined,
+}
+
+/// Sorted order (either Ascending or Descending).
+/// To get a regular top-K results search, use `SortOrder::Desc`.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum SortOrder {
+    /// Descending. This is the default to get Top-K results.
+    Desc,
+    /// Ascending order.
+    Asc,
+}
+
+/// The `SortingFieldComputer` can be seen as the specialization of `SortBy` applied to a specific `SegmentReader`.
+/// Its role is to compute the sorting field given a `DocId`.
+enum SortingFieldComputer {
+    SortByFastField {
+        fast_field_reader: DynamicFastFieldReader<u64>,
+        order: SortOrder,
+    },
+    /// If undefined, we simply sort by DocIds.
+    Undefined,
+}
+
+impl SortingFieldComputer {
+    /// Returns the ranking key for the given element
+    fn compute_sorting_field(&self, doc: DocId) -> u64 {
+        match self {
+            SortingFieldComputer::SortByFastField {
+                fast_field_reader,
+                order,
+            } => {
+                let field_val = fast_field_reader.get(doc);
+                match order {
+                    // Descending is our most common case.
+                    SortOrder::Desc => field_val,
+                    // We get Ascending order by using a decreasing mapping over u64 as the sorting_field.
+                    SortOrder::Asc => u64::MAX - field_val,
+                }
+            }
+            SortingFieldComputer::Undefined => 0u64,
+        }
+    }
+}
+
+fn resolve_sort_by(
+    sort_by: &SortBy,
+    segment_reader: &SegmentReader,
+) -> tantivy::Result<SortingFieldComputer> {
+    match sort_by {
+        SortBy::SortByFastField { field_name, order } => {
+            if let Some(field) = segment_reader.schema().get_field(&field_name) {
+                let fast_field_reader = segment_reader.fast_fields().u64_lenient(field)?;
+                Ok(SortingFieldComputer::SortByFastField {
+                    fast_field_reader,
+                    order: *order,
+                })
+            } else {
+                Ok(SortingFieldComputer::Undefined)
+            }
+        }
+        SortBy::Undefined => Ok(SortingFieldComputer::Undefined),
+    }
+}
+
+/// PartialHitHeapItem order is the inverse of the natural order
+/// so that we actually have a min-heap.
+#[derive(Clone, Copy)]
+struct PartialHitHeapItem {
+    sorting_field: u64,
+    doc: DocId,
```

Comment:
sometimes we have `doc`, sometimes we have `doc_id` , would be better to have doc_id everywhere

Reply:
also I was a little bit confused about `sorting_field`, I thought at the beginning that it was the field sorting id! Either document it or call it `sorting_field_value` ?

Reply:
i changed to doc_id everywhere.

Reply:
renamed sorting_field_value